### PR TITLE
Fix memory allocation error when no. of device config entries is zero

### PR DESCRIPTION
**Pull request by deepsworld**
*Description of changes:*
Added a conditional check to allocate memory for device configuration list in provisioning parser only if the 
configuration entries is greater than zero. 

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

**Comment by aggarw13** [comment]
Hi @deepsworld, can you share whether you encountered an issue with the Fleet Provisioning SDK?

Regarding your proposed changes in the PR, if an entry for Device Configuration has been found in the response payload (from the `RegisterThing` API) by the decoder, the number of entries must be greater than zero; otherwise, it is an invalid CBOR syntax. Instead of a having a check in conditional form, it can be an assert check to detect decoder issues during development.

**Comment by deepsworld** [comment]
Hey @aggarw13, 
Sorry I should have raised an issue first. Below are the logs for provisioning when there is no device configuration attributes. 
`[INFO ][MQTT][2021-03-30 15:06:11] (MQTT connection 0xa94578) SUBSCRIBE operation scheduled.
[INFO ][MQTT][2021-03-30 15:06:11] (MQTT connection 0xa94578, SUBSCRIBE operation 0xa947d8) Waiting for operation completion.
[INFO ][MQTT][2021-03-30 15:06:11] (MQTT connection 0xa94578, SUBSCRIBE operation 0xa947d8) Wait complete with result SUCCESS.
[INFO ][MQTT][2021-03-30 15:06:11] (MQTT connection 0xa94578) SUBSCRIBE operation scheduled.
[INFO ][MQTT][2021-03-30 15:06:11] (MQTT connection 0xa94578, SUBSCRIBE operation 0xa947d8) Waiting for operation completion.
[INFO ][MQTT][2021-03-30 15:06:11] (MQTT connection 0xa94578, SUBSCRIBE operation 0xa947d8) Wait complete with result SUCCESS.
[ERROR][Provisioning][2021-03-30 15:06:12] Failure in allocating memory for device configuration data in response payload of REGISTER DEVICE operation
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) UNSUBSCRIBE operation scheduled.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578, UNSUBSCRIBE operation 0xa94c48) Waiting for operation completion.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578, UNSUBSCRIBE operation 0xa94c48) Wait complete with result SUCCESS.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) UNSUBSCRIBE operation scheduled.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578, UNSUBSCRIBE operation 0xa94c48) Waiting for operation completion.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578, UNSUBSCRIBE operation 0xa94c48) Wait complete with result SUCCESS.
**[ERROR][IOT_CLIENT][2021-03-30 15:06:12] Failed to provision client application, error NO MEMORY**
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) Disconnecting connection.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578, DISCONNECT operation 0xa94c48) Waiting for operation completion.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578, DISCONNECT operation 0xa94c48) Wait complete with result SUCCESS.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) Connection disconnected.
[INFO ][NET][2021-03-30 15:06:12] Shutting down TLS connection.
[ERROR][NET][2021-03-30 15:06:12] Error receiving from socket 3.
[WARN ][NET][2021-03-30 15:06:12] Receive requested 1 bytes, but only 0 bytes received.
[ERROR][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) Unknown packet type ff received.
[ERROR][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) Error processing incoming data. Closing connection.
[INFO ][NET][2021-03-30 15:06:12] Shutting down TLS connection.
[INFO ][NET][2021-03-30 15:06:12] TLS connection closed.
[INFO ][NET][2021-03-30 15:06:12] Connection (socket 3) shut down.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) Network connection closed.
[ERROR][NET][2021-03-30 15:06:12] Detected error on socket 3, receive thread exiting.
[INFO ][NET][2021-03-30 15:06:12] TLS connection closed.
[INFO ][NET][2021-03-30 15:06:12] Connection (socket 3) shut down.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) Network connection closed.
[INFO ][NET][2021-03-30 15:06:12] (Socket 3) Socket closed.
[INFO ][MQTT][2021-03-30 15:06:12] (MQTT connection 0xa94578) Network connection destroyed.
[INFO ][Provisioning][2021-03-30 15:06:12] api: Completed cleanup of library.
[INFO ][MQTT][2021-03-30 15:06:12] MQTT library cleanup done.
[INFO ][NET][2021-03-30 15:06:12] Network library cleanup done.
[INFO ][INIT][2021-03-30 15:06:12] SDK cleanup done.`

**Comment by aggarw13** [comment]
Hi @deepsworld, sorry for the late response, and thank you for sharing the case about how you are encountering zero entries in the map, which makes sense to me if you have defined no entries for the `DeviceConfiguration` field in the Provisioning template.  
Therefore, your proposed changes look good to me 🙂

**Review by aggarw13 — APPROVED** [review]

File: libraries/aws/provisioning/src/aws_iot_provisioning_parser.c [hunk 1]
```diff
@@ -629,17 +629,20 @@ AwsIotProvisioningError_t _AwsIotProvisioning_ParseRegisterThingResponse( AwsIot
...
+                     pDeviceConfigurationList = AwsIotProvisioning_MallocDeviceConfigurationList( numOfDeviceConfigurationEntries *
                                                                                              sizeof( AwsIotProvisioningResponseDeviceConfigurationEntry_t ) );
 
-                if( pDeviceConfigurationList == NULL )
-                {
-                    IotLogError( "Failure in allocating memory for device configuration data in response payload of %s operation",
+                    if( pDeviceConfigurationList == NULL )
+                    {
+                        IotLogError( "Failure in allocating memory for device configuration data in response payload of %s operation",
                                  REGISTER_THING_OPERATION_LOG );
```
Comment (archigup) [inline libraries/aws/provisioning/src/aws_iot_provisioning_parser.c:639]:
```suggestion
                                     REGISTER_THING_OPERATION_LOG );
```

File: libraries/aws/provisioning/src/aws_iot_provisioning_parser.c [hunk 2]
```diff
@@ -629,17 +629,20 @@ AwsIotProvisioningError_t _AwsIotProvisioning_ParseRegisterThingResponse( AwsIot
                     IOT_SET_AND_GOTO_CLEANUP( AWS_IOT_PROVISIONING_INTERNAL_FAILURE );
                 }
 
-                pDeviceConfigurationList = AwsIotProvisioning_MallocDeviceConfigurationList( numOfDeviceConfigurationEntries *
+                if ( numOfDeviceConfigurationEntries > 0) {
+                     pDeviceConfigurationList = AwsIotProvisioning_MallocDeviceConfigurationList( numOfDeviceConfigurationEntries *
                                                                                              sizeof( AwsIotProvisioningResponseDeviceConfigurationEntry_t ) );
```
Comment (archigup) [inline libraries/aws/provisioning/src/aws_iot_provisioning_parser.c:634]:
```suggestion
                if( numOfDeviceConfigurationEntries > 0 )
                {
                    pDeviceConfigurationList = AwsIotProvisioning_MallocDeviceConfigurationList( numOfDeviceConfigurationEntries *
                                                                                                 sizeof( AwsIotProvisioningResponseDeviceConfigurationEntry_t ) );
```

**Review by archigup — CHANGES_REQUESTED** [review]
The formatting needs a few minor changes.

File: libraries/aws/provisioning/src/aws_iot_provisioning_parser.c (hunk 1 above)
Comment (deepsworld) [reply libraries/aws/provisioning/src/aws_iot_provisioning_parser.c:639]: @archigup I am not sure what to do here! Do I just select `commit suggestion` to make the changes?

Comment (archigup) [reply libraries/aws/provisioning/src/aws_iot_provisioning_parser.c:639]: @deepsworld Yes, you can use `commit suggestion` to add the changes, and that will apply the formatting changes. Thanks!

**Review by archigup — APPROVED** [review]

**Review by n9wxu — APPROVED** [review]

**Pull request merged**